drivers/include/gpio_exp.h

```diff
@@ -0,0 +1,320 @@
+                   gpio_exp_pin(pin), mode, flank, cb, arg); \
+        } \
+    } \
+    while (0)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_irq_enable calls
+ */
+#define GPIO_INTERCEPT_IRQ_ENABLE(pin) \
+    do { \
+        if (pin > GPIO_EXP_THRESH) { \
+            gpio_exp_t *exp_entry = gpio_exp_entry(pin); \
+            if (exp_entry == NULL) { \
+                return; \
+            } \
+            exp_entry->driver->irq(exp_entry->dev, gpio_exp_pin(pin), 1); \
+            return; \
+        } \
+    } \
+    while (0)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_irq_disable calls
+ */
+#define GPIO_INTERCEPT_IRQ_DISABLE(pin) \
+    do { \
+        if (pin > GPIO_EXP_THRESH) { \
+            gpio_exp_t *exp_entry = gpio_exp_entry(pin); \
+            if (exp_entry == NULL) { \
+                return; \
+            } \
+            exp_entry->driver->irq(exp_entry->dev, gpio_exp_pin(pin), 0); \
+            return; \
+        } \
+    } \
+    while (0)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_read calls
+ */
+#define GPIO_INTERCEPT_READ(pin) \
+    do { \
+        if (pin > GPIO_EXP_THRESH) { \
+            gpio_exp_t *exp_entry = gpio_exp_entry(pin); \
+            if (exp_entry == NULL) { \
+                return -1; \
+            } \
+            return exp_entry->driver->read(exp_entry->dev, gpio_exp_pin(pin)); \
+        } \
+    } \
+    while (0)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_set calls
+ */
+#define GPIO_INTERCEPT_SET(pin)    GPIO_INTERCEPT_WRITE(pin, 1)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_clear calls
+ */
+#define GPIO_INTERCEPT_CLEAR(pin)    GPIO_INTERCEPT_WRITE(pin, 0)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_toggle calls
+ */
+#define GPIO_INTERCEPT_TOGGLE(pin) \
+    do { \
+        if (pin > GPIO_EXP_THRESH) { \
+            gpio_exp_t *exp_entry = gpio_exp_entry(pin); \
+            if (exp_entry == NULL) { \
+                return; \
+            } \
+            if (exp_entry->driver->read(exp_entry->dev, gpio_exp_pin(pin))) { \
+                exp_entry->driver->write(exp_entry->dev, \
+                                         gpio_exp_pin(pin), 0); \
+            } \
+            else { \
+                exp_entry->driver->write(exp_entry->dev, \
+                                         gpio_exp_pin(pin), 1); \
+            } \
+        } \
+    } \
+    while (0)
+
+/**
+ * @brief   Macro for intercepting and redirecting gpio_write calls
+ */
+#define GPIO_INTERCEPT_WRITE(pin, value) \
+    do { \
+        if (pin > GPIO_EXP_THRESH) { \
+            gpio_exp_t *exp_entry = gpio_exp_entry(pin); \
+            if (exp_entry == NULL) { \
+                return; \
+            } \
+            exp_entry->driver->write(exp_entry->dev, \
+                                     gpio_exp_pin(pin), value); \
+        } \
+    } \
+    while (0)
```

Comment:
All this macro substitution should be in a static inline function, then handle the return code (if any) into the caller function. This way it's clear that this code is terminating the routine.

Reply:
Unfortunately it seems that this is a case where it is impossible to both be clear and be concise. To use inline functions would require:
1) an ```#ifdef``` to check if the module is enabled
2) an ```if``` statement to check if the pin needs redirecting
3) the redirection
4) a return, if required by either redirection or error

In order to have all of this logic, it would require a minimum of 5-6 lines (depending on whether the return needs to be on a separate line or not). I can add this logic, but unfortunately it would be duplicated for every macro instance. I feel that due to the duplication it is more appropriate to use a macro in this case. The macro incorporates all of the above logic.

Reply:
Removed and now uses separate functions for the low-level implementation.